In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw10.ipynb")

# Final Project: Predictive Modelling and Model Comparision

Names: Aaryabrat Chhatkuli, Adip Shrestha

Student IDs: 605097, 603619
Collaborators: Course Zybook


## Instructions

This homework is an _open-ended final project_ for our class.

We will focus on Data Profiling and Supervise Machine Learning including Feature Engineering, Feature Selection, and Model Selection & Comparison.

We will be using the techniques and approaches introduced throughout the course focusing on the following topics:

* [M0] Python for DS
* [M1] The DS and ML Workflows
* [M2] Exploratory Data Analsyis
* **[M3] Linear Regression (if applicable - depends on problem/dataset)**
* **[M4] Linear Classification (if applicable - depends on problem/dataset)**
* [M5] Sentiment Analysis (if applicable - depends on problem/dataset)
* [M6] Ethical Thinking for Data Science (if applicable - depends on problem/dataset)
* **[M7] Performance Evaluation: Overfitting & Cross-Validation**
* **[M8] Clustering K-Means (if applicable - depends on problem/dataset)** 
* **[M9] Similarity-based Learning K-NN (if applicable - depends on problem/dataset)**
* **[M10] More Models (Decision Tree, Random Forest, Neural Network) for Feature Engineering & Model Comparision & Selection**
* **[M11] DS Workflow in Python: Data Profiling and ML Pipelines**

The main focus of this project is majorly on the last modules [M7]-[M11]. In particular, it will be helpful to review **Lab 3 & hw 3 onwards for more clarification**.
Most of the things you will do in this final project are explained in `Lab10` and `Lab11`. Then, review the other labs and hws to recap the general workflow of our DS approaches to the various problems we have worked on so far. This will help you to be prepared for all the steps in this project, so that you do not miss anything. 

In general, you should feel free to import any package that we have previously used in class. Ensure that all plots have the necessary components that a plot should have (e.g. axes labels, a title, a legend).

> **!! Do not use any packages/methods/models that we did not discuss in this class !!**

Additionally, note that this assignment is more like a project than a typical homework and it will have a slightly different structure than the hws in the past. One of the reasons why Jupyter Notebooks are so popular in the field is because you can really treat them like notebooks to explain your findings as you discover them. In this assignment, we provide some **Problem** stubs following the DS workflow and you will need to fill in the analysis and structure as you go.

Furthermore, in addition to recording your collaborators on this homework, please also remember to cite/indicate all external sources used when finishing this assignment. This includes peers, TAs, and links to online sources. 

Frequently **save** your notebook!

### Collaborators and Sources
Furthermore, in addition to recording your **collaborators** on this homework, please also remember to **cite/indicate all external sources** used when finishing this assignment. 
> This includes peers, TAs, and links to online sources. 

Note that these citations will not free you from your obligation to submit your _own_ code and write-ups, however, they will be taken into account during the grading and regrading process.

In [2]:
# collaborators and sources:
# Albert Einstein and Marie Curie
# https://developers.google.com/edu/python/strings

# your code here
answer = 'my answer'

### Submission instructions
* Submit this Python notebook, including your answers in the code cells as homework submission.
* **Feel free to add as many cells as you need to** — just make sure you don't change what we gave you. 
* **Does it spark joy?** Note that you will be partially graded on the presentation (_cleanliness, clarity, comments_) of your notebook so make sure you [Marie Kondo](https://lifehacker.com/marie-kondo-is-not-a-verb-1833373654) your notebook before submitting it.

In [3]:
import os
os.environ["OMP_NUM_THREADS"] = "1" # run this first!

## 1. Introduction

For our final project this semester, we'll take a look at data for Melbourne Housing Price Prediction
https://www.kaggle.com/datasets/dansbecker/melbourne-housing-snapshot.


As you have seen throughout the semseter we have learned various techniques to solve a problem task (Classification, Regression, Clustering etc.) starting from problem formulation to EDA to model building to model evaluation and performing feaure engineering and selection in order to improve the model's performance.
You will be combining all the knowledge learned so far in this final project. You will be making assuptions and decisions on your own with appropriate justifications in writeups about how to improve the prediction model using all the tools and techniques learned so far and provide the best model at the end.


The project aims at building a model of housing prices to predict median house values in Melbourne using the provided dataset. This model should learn from the data and be able to predict the median housing price in any district, given all the other metrics.

Our goal will be to use this dataset to gain some insight about characteristics of different features.
We will be using data profiling from pandas and create models (hopefully more than 3) and compare them to find out which model performed the best and probably explain why it is the best model among all in this data.


### Problem 1.1

Download the data and load it in this notebook for further processing and analysis.

In [4]:
# your code here
# Note: Imports to run the code
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

data = pd.read_csv('./data/melb_data.csv')
data

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,...,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,3067.0,...,1.0,1.0,202.0,NaN,NaN,Yarra,-37.79960,144.99840,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,3067.0,...,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.80790,144.99340,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,3067.0,...,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.80930,144.99440,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,3067.0,...,2.0,1.0,94.0,NaN,NaN,Yarra,-37.79690,144.99690,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,3067.0,...,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.80720,144.99410,Northern Metropolitan,4019.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13575,Wheelers Hill,12 Strada Cr,4,h,1245000.0,S,Barry,26/08/2017,16.7,3150.0,...,2.0,2.0,652.0,NaN,1981.0,NaN,-37.90562,145.16761,South-Eastern Metropolitan,7392.0
13576,Williamstown,77 Merrett Dr,3,h,1031000.0,SP,Williams,26/08/2017,6.8,3016.0,...,2.0,2.0,333.0,133.0,1995.0,NaN,-37.85927,144.87904,Western Metropolitan,6380.0
13577,Williamstown,83 Power St,3,h,1170000.0,S,Raine,26/08/2017,6.8,3016.0,...,2.0,4.0,436.0,NaN,1997.0,NaN,-37.85274,144.88738,Western Metropolitan,6380.0
13578,Williamstown,96 Verdon St,4,h,2500000.0,PI,Sweeney,26/08/2017,6.8,3016.0,...,1.0,5.0,866.0,157.0,1920.0,NaN,-37.85908,144.89299,Western Metropolitan,6380.0


**Write-up!** Describe the data, answering questions including, but not limited to, these: Where does the data come from? How was it obtained? How many examples and features does the dataset have? What kinds of features are in the dataset? What values can these features take? What kind of data science problem is this? What application area does the dataset fall into? What kind of prediction (regression, classification or clustering) we are doing with this dataset and why?
> **Hint**: Recall, our goal is to build a model of housing prices to predict median house values.

**Your Answer Here:**

This is a Melbourne Housing Dataset, the content of which is a snapshot of a datset created by Tony Pino and it is published on Kaggle. The main data of which this is a snapshot of was scraped from publicly available results posted every week from Domain.com.ai. According to what is written on the main Melbourne Housing Market Dataset and the snapshot , "The dataset includes Address, Type of Real estate, Suburb, Method of Selling, Rooms, Price, Real Estate Agent, Date of Sale and distance from C.B.D.". The snapshot we are using is a static snapshot of the Melbourne Housing Market Dataset that was created in September 2017, and in this one home with no Price have been removed.

After loading this dataset and just running data at the end, we can see that the size of the data is 13580 rows X 21 columns, where each row is for one house property basically. The columns have a mix of numerical and categorical features with geographic features also within them. A breakdown of the features are, as is on the Kaggle description and looking at the CSV file:

***Numerical Features (Continuous)***

Distance: Distance from CBD

Landsize: Land Size

BuildingArea: Building Size

Lattitude: Geographical latitude of the house

Longitude: Geographical longitude of the house

Propertycount: Number of properties that exist in the suburb.


***Numerical Features (Discrete)***

Rooms: Number of rooms

Bedroom2 : Scraped # of Bedrooms (from different source)

Bathroom: Number of Bathrooms

Car: Number of carspots

YearBuilt: Year the house was built

***Categorical Features (Nominal)***

Suburb: Suburb where the property is located

Postcode: Postcode of the house, although numberical value, the feature itself is cateogrical

Address: Address of the house

Type: PropeType: br - bedroom(s); h - house,cottage,villa, semi,terrace; u - unit, duplex; t - townhouse; dev site - development site; o res - other residential.
rty Type

Method: S - property sold; SP - property sold prior; PI - property passed in; PN - sold prior not disclosed; SN - sold not disclosed; NB - no bid; VB - vendor bid; W - withdrawn prior to auction; SA - sold after auction; SS - sold after auction price not disclosed. N/A - price or highest bid not available.

SellerG: Real Estate Agent

Regionname: General Region (West, North West, North, North east …etc)

CouncilArea: Governing council for the area

***Temporal Feature (related to time that's why)***

Date: Date sold

***Target Variable***

Price: Price in dollars

The project is basically a form of analysis on real-estate data, covering the concepts of modeling different parts of the housing market to predict price. Since the dataset has labeled examples like each property is associated with a sale price, from a machine learning perspective this is a supervised learning problem, where the main core idea as stated earlier is to predict the continuous numerical value of the house price, make this a regression focused task. So the goal here is to build a model that goes through and learns from the information in the dataset and predict the median house values in Melbourne. 

***Reference:*** https://www.kaggle.com/datasets/dansbecker/melbourne-housing-snapshot/data
***Reference:*** https://www.kaggle.com/datasets/anthonypino/melbourne-housing-market

### Problem 1.2: Getting Familiar with the Data

Generate the data report using __ProfileReport__ as we did in lab 11.

In [6]:
# your code here
# Note: First let us import the ProfileReport module
from ydata_profiling import ProfileReport

# Note: Next we will generate the profile report following similar style as Lab 11
profile = ProfileReport(
    data,
    title = 'Melbourne Housing Data Report using ProfileReport',
    correlations = {
        'pearson': {'calculate': True},
        'spearman': {'calculate': False},
        'kendall': {'calculate': False},
        'phi-k': {'calculate': False},
        'cramers': {'calculate': False}
    }
)
profile

# Reference: https://docs.profiling.ydata.ai/latest/

AttributeError: module 'numba' has no attribute 'generated_jit'

### Problem 1.3: Data Description

**Write-up!** What are the major highlights and interesting aspects in the report generated in the previous cell (or in your own EDA process). 
Describe your insights using the data/features/visualizations/statistics. Be sure to consider the relationship among features such as correlation and interaction, as well as distributions, missing data, outliers. Include additional visualizations as necessary.

In [7]:
# your code here



**Answer here:** (feel free to make multiple cells!)

### Problem 1.4: Split the Data

Extract input (X) and output (y) data from the dataset. 

Use the train and test split from the scikit learn library to split the data into 80% training and 20% testing.

Then, initalize the Kfolds. Be sure to use these folds when executing cross validation below so that the results are comparable.

In [8]:
# your code here
from sklearn.model_selection import train_test_split, KFold

# Note: First we are going to extract the input variable X and the target variable y
X = data.drop(columns = ['Price'])
y = data['Price']

# Note: Next step is to actually split the data into the 80/20 train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

# Note: As said in the question, we will also initialize the KFolds, for now we will use 5 as it is around standard but later on might change depending on how we
# progress, ofc avoiding any form of data snooping though :)
kfolds = KFold(n_splits=5, shuffle=True, random_state=42)

# Note: Let us now display the information we obtain with the above split
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((10864, 20), (2716, 20), (10864,), (2716,))

**Write-up!** Why 80% train and 20% test split is considered a good practice? How would you pick the train-test split size if the dataset is relatively small or large? What is the purpose of using cross validation technique and how does it ensure model robustness? What is Data snooping and how do you avoid it?

**Answer here:** (feel free to make multiple cells!)

80/20 split is much standard in the sense of splitting and it has its reason on why it is considered a good practice. Mainly, this is becasue it gives a good balance between how much training data you use where 80% of the data being used for training makes sure the model does learn the patterns as we want it to learn, while at the same time not so much so as to cause overfitting where the 20% split comes in and this provides a good amount of data to test on, because if we think about the sizes of datasets, datasets are large with thousands of values, and even in this case if we look, like the test datset has 2716 values, meaning that such portions while they look small are actually large because 20% of thousands is also thousands of data values. And so this way, we use this 80/20 split because it gives much training data to learn enough information, while using the test as 20% gives enough on the test set, and is relevant to evaluate model performance, and this balance helps avoid overfitting and underfitting, as described, and allows the model to generalize well.

Like mentioned earlier, in this case with the 20% split itself we have 2700+ data points on the test sets. Buit realistically if we are comparing lets take a small dataset as having like less than 1500-1000 rows then it would make more sense to use more training data as a 90/10 split or even more depending on how low we are going because the model needs as many data examples as possible here to properly learn. If we are looking at tens of thousands as a large dataset, obviously bigger than this but this could itself be considered, then smaller test sets are still large like we saw here so like 80/20 or 85/15 does make sense, even 70/30 as have used that for projects of this scale before too, so yeah that is a valid split. If we are looking at extremely large datasets where we are touching the hundreds of thousands, even millions then percentage is less relevant there, because you are going to have a lot of data points to train and test, so maybe still having the majority as the train dataset would be more ideal.

The purpose behind CV has a lot of answers, but to first describe the why lets look at the what. Using Cv divides the training data into different folds, where each fold you are training on some and validating on the rest, and this way try to cover all the data points as atleast once being in the Validation and in the training. By doing this, we can make sure that the model gets evaluated on many different subsets, and this makes sure that rather than just using a fixed validation set and getting a performance that otherwise could be considered a statistical chance of "luck" to say, this avoids such misconceptions or well reduces such ideas. At the same time, it makes it more sensible when we compare it with the test set, where it helps explain the performance better understanding the concepts of overfitting and underfitting looking at test set performance vs CV performance. We are basically comparing the model on data it hasn't seen before here too, and of course the idea behind how validation itself works. CV also has important role in hyperparameter tuning by looking at different performances for different hyperparameters, like the depth in RF or number of neighbors in KNN as examples.

Data snooping also called data leakage is a situation when information from the test set basically leaks into the training process. It's almost like looking at the test data, and then fitting ur training phase as such to get a good test score. Considering it looking at or knowing the exam questions while you are studying so you know what to study. This causes our train-test evaluation to be good, but in the real world scenarios the model would otherwise perform poorly, again because of aforementioned reasons. Other examples more in this ML side would be like if we scale the dataset use the entire dataset before splitting, or use teh test set outcomes to then control feature engineering, or looking at the scores in the test-set and adjusting our model performance. Ways to avoid it would be to firstly split the data first hand, then perform the preprocessing steps only on the training set, and keeping the test data literally isolated until the final model performance evaluation. Another way is to also utilize pipelines in the process to avoid any form of leakage.

## Problem 2: Preprocessing and Baseline Models

### Problem 2.1: Preprocessing for the Baseline Models

Let's start by preprocessing the data so that we can train some baseline models. Note that categorical variables will need to be encoded and missing values should be appropriately delt with using imputation. Also, some models may require scaling. Imputation and scaling should be done in a manner that aovids data snooping.
> **Hint:** The use of Scikit-Learn's Pipelines are highly encouraged!

> **Hint:** Not all columns in the provided dataset can or should be used as features directly. If you see any features that need feature extraction to be useful or that don't make any sense to be used, remove them and **document/justify your choices**. Go back to features left/marked for feature extraction in Pronlem 3. 

In [9]:
# your code here
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer

# Note: here we are separating input (X) and output (y)
X = data.drop(columns=['Price'])
y = data['Price']

# Note: Now we would like to indentfy the numerical and categorial columns
numerical_features = X.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X.select_dtypes(include=['object']).columns

#Note: We would no do numerical preprocessing whih is imputing the missing values and scaling it. 
numerical_transformer = Pipeline(steps=[('imputer', SimpleImputer(strategy='median')),('scaler', StandardScaler())])

# Note: here we are doing the categorical preprocessing which is to impute the missig values and using encoder; one hot encode 
categorical_transformer = Pipeline(steps=[('imputer', SimpleImputer(strategy='most_frequent')),('encoder', OneHotEncoder(handle_unknown='ignore'))])

# Note: Now Combining transformations
preprocessor = ColumnTransformer(transformers=[('num', numerical_transformer, numerical_features),('cat', categorical_transformer, categorical_features)])

# Note: We would like to fit only on the training data which is to avoid data snooping or leakage
preprocessor.fit(X_train)

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('scaler', StandardScaler())]),
                                 Index(['Rooms', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude',
       'Propertycount'],
      dtype='object')),
                                ('cat',
                                 Pipeline(steps=[('imputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('encoder',
                                                  OneHotEncoder(handle_unknown='ignore'))]),
                                 Index(['Suburb', 'Address', 'Type', 'Method', 'SellerG', 'Date', 'CouncilArea',
       'Regionname'],
      dtype='object'))])

### Problem 2.2: Perform Linear Regression

Perform Linear Regression on training data. Predict the output for the test dataset using the fitted model. Print the root mean squared error (RMSE) and R^2 value from Linear Regression for test data set.


In [10]:
# your code here
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Note: We are creating a full pipe line which is connecting preprocessing and the model
linear_model = Pipeline(steps=[('preprocessor', preprocessor),('model', LinearRegression())])

#Note: next step would be to train the model on our training data 
linear_model.fit(X_train, y_train)

# Note: predicting on the test set 
y_pred = linear_model.predict(X_test)

# Note: we would use that prediction to Compute metrics
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("Linear Regression Results:")
print(f"RMSE: {rmse:.4f}")
print(f"R^2: {r2:.4f}")



Linear Regression Results:
RMSE: 505858.3593
R^2: 0.3558


**Write-up!** What does R^2 tell us about the model's performance? What does RMSE tell us about the model's performance?

**Answer here:** (feel free to make multiple cells!)

The R^2 value of 0.3558 means that out linear regression model here explains about 35% of the variation in our house prices. This is actually slightly dissapointing. This would be interpreted as if we were trying to predict the price of the housing using this model we would be accounting for only around third of the whole story. the remaininf two thirds which would still be an unknown. This means that the real world of house prices is so much messier than a straight line could explain, also giving us an idea a more complex, non-linear realtionship must be at play here. There must be a better model to use for price prediction.

Now lets take a look at our value for RMSE which is 505858.3593
That is a big number indicating a huge margin of error. On average our prediction is off by around hald a million which in housing prices is pretty big and frankly speraking its not acceptable to make real estate decisions. From our R^2 value and RMSE value we can clearly see that the prtices are not properly fit by a linear model and the data is non linear. This was a good starting point but now we need better models like KNN decision trees or random forests that can handle such twist and turns in the data.

### Problem 2.3 : Perform K-Nearest Neighbors Regression

Perform K-Nearest Neighbors Algorithm for Regression using cross validation to tune any hyperparameters, such as `n_neighbors`.

Then, using the optimal hyperparameters, train a K-Nearest Neighbors Regressor on the entirety of the training data and predict output for the test dataset using the fitted model.

Print root mean squared error (RMSE) and R^2 from K-Nearest Neighbors Regression on the test datset.

In [11]:
# your code here
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Note : Step 1 here would be to again as above build  a full pipeline from preprocessing to model
knn_pipeline = Pipeline(steps=[('preprocessor', preprocessor),('model', KNeighborsRegressor())])

# Note: next step is to tune n_neighbour using crossvalidation
n_neighbours = {'model__n_neighbors': [3, 5, 7, 9, 11, 15, 20]}

knn_grid_for_cv = GridSearchCV(knn_pipeline,n_neighbours,cv=kfolds,scoring="neg_root_mean_squared_error",n_jobs=-1)

knn_grid_for_cv.fit(X_train, y_train)

print("The best KNN hyperparameters:", knn_grid_for_cv.best_params_)

# Note: finally training the final model and evaluating it on our test set
y_pred = knn_grid_for_cv.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("KNN Regression Results:")
print("RMSE:", rmse)
print("R^2:", r2)



The best KNN hyperparameters: {'model__n_neighbors': 15}
KNN Regression Results:
RMSE: 322067.1849766118
R^2: 0.7388629057593333


**Write-up!** How K-NN is different than Linear regression model and what additional steps you have to take for KNN Regression?

**Answer here:** (feel free to make multiple cells!)

The fundamental difference: linear regression is like drawing a best fiot straight line through our data and saying it done. Linear regression assumes that opur data is linear it learns a fixed set of coefficients, and then uses that same equation for every prediction. there is a lot of benefit for this methjod it is fast it is simple it is eas yt opunderstadnd but at the same time it is very rigid model. Like in our dataset for melbourne houses if the datga is nopn linear we are  out of luckl. K-NN in the other hand doesnt have such rules, it is a non parametric model it memorizes the entire training set and makes predictions by looking at the k nearest neighbors to whatever house you're trying to price. The model averages the K neighbours values and tells us well look there are this list of similar houses like yours and they cost around this  X value so your house must also cost X. This local, flexible approach makes K-NN far better at capturing the weird, nonlinear patterns that exist in real housing markets. This is why we can also look at opurt results and see K-NN beat linear by a huge difference.

With this model there are some extra work that needed to be done :
Since our model is very flexible we have some extra hoops we need to jump for it to work. Sinc K-NN relies entirely in measuring distances between data points, there arises a critical step which is scaling our features. Without this step features that have very large values can totally dominate the predictions and overshadow smaller features. For example the land size would make smaller features like number of bathrooms obsolete. This is why we used Standard Scaler in our preprocessing to handle such situations. After this we have another step which is to choose K, the number of K to use here we use grid search CV to find k value. The downside I would think is that it is slower than linear regression and computationally expensive.

### Problem 2.4 : Perform Decision Tree Regression 

Perform Decision Tree Regression using cross validation to tune any hyperparameters, such as `max_depth`.

Then, using the optimal hyperparameters, train a Decision Tree Regressor on the entirety of the training data and predict output for the test dataset using the fitted model.

Print root mean squared error (RMSE) and R^2 from Decision Tree Regression on the test dataset.

In [12]:
# Note: We are importing the Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor

# Note: Here we would build a pipeline that includes from preprocessing to the decision tree model 
dec_tree_pipeline = Pipeline(steps=[ ('preprocessor', preprocessor), ('model', DecisionTreeRegressor(random_state=42))])

# Note: we are building a hyperparameter grid for runing now. Where - 
# max_depth controls how deep the tree can grow
# min_samples_split controls how many samples needed to split a node
# min_samples_leaf controls how many samples minimum per leaf node
dec_tree_hyperparam = {'model__max_depth': [3, 5, 7, 10, 15, 20, None],'model__min_samples_split': [2, 5, 10],'model__min_samples_leaf': [1, 2, 4]}

# Note: Now we do GridSearchCV which performs cross-validation on each combination
dec_tree_grid = GridSearchCV( dec_tree_pipeline,dec_tree_hyperparam, cv=kfolds,scoring="neg_root_mean_squared_error", n_jobs=-1)

# Note: Next again as needed Fitting the model on the training data and we use CV to find the best parameters
dec_tree_grid.fit(X_train, y_train)

# Note: We are just printing the best hyperparameters found
print("Best Decision Tree hyperparameters:", dec_tree_grid.best_params_)

# Note: Finally Predicting on the test set using the best model trained on the training data
y_pred = dec_tree_grid.predict(X_test)

# Note: using the prediction to compute evaluation metrics
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)


print("Decision Tree Regression Results:")
print("RMSE:", rmse)
print("R^2:", r2)



Best Decision Tree hyperparameters: {'model__max_depth': 10, 'model__min_samples_leaf': 4, 'model__min_samples_split': 10}
Decision Tree Regression Results:
RMSE: 335963.30664063874
R^2: 0.7158423799722311


**Write-up!** What are the advantages or disadvantages of using Decision Tree over Linear Regression and KNN?

**Answer here:** (feel free to make multiple cells!)
Okay, so the fascinating part comes with Decision Trees. Decision Trees say, "Yeah, the world is messy and nonlinear let's deal with it," in contrast to Linear Regression, which stubbornly maintains that everything follows a straight line. In order to capture all those strange, complex relationships that genuinely occur in real estate markets, they divided the data into branches based on yes-or-no questions. The best part is that they don't require feature scaling. we can recall how K-NN gets upset if we dont standardize everything.Decision trees don't really care about that. Preprocessing becomes much less of a hassle because they are merely creating divisions based on thresholds rather than computing distances, regardless of the size of the numbers. Additionally, they are remarkably simple to understand and handle mixed data types with grace. You can actually sketch out the tree, trace its branches, and see just how it determined that a house should cost X dollars.Decision tree results  R2 of 0.716 and an RMSE of almost $336,000, the Decision Tree completely outperformed linear regression and is super close to K-NN model.

However, this is the point at which I must somewhat shatter the bubble: If you allow decision trees run recklessly, they become overfitting machines. Without any safeguards, they will learn every characteristic in the training set : every strange outlier, every spike in random noise—and then become absolutely stuck when they come across new homes. It's similar to studying for an exam by remembering every word of the practice test rather than truly understanding the material. We had to use GridSearchCV to tune it out completely because of this. Without tuning parameters like max_depth, min_samples_split, and min_samples_leaf, the model can perform extremely well on training data but generalize poorly on unseen data. And unlike K-NN's smooth averages, Decision Tree predictions are chunky, they jump abruptly between values. Still, despite these characteristics, they're fast, interpretable, and flexible enough to handle the chaos of real-world data.

### Problem 2.5 : Perform Random Forest Regression

Perform Random Forest Regression using cross validation to tune any hyperparameters, such as `n_estimators`.

Then, using the optimal hyperparameters, train a Random Forest Regressor on the entirety of the training data and predict output for the test dataset using the fitted model.

Print root mean squared error (RMSE) and R^2 from Random Forest Regression on the test dataset.

In [13]:
# your code here
from sklearn.ensemble import RandomForestRegressor

# Note: Again building a pipeline from preprocessing to model
ran_for_pipeline = Pipeline(steps=[('preprocessor', preprocessor),('model', RandomForestRegressor(random_state=42))])

# Note: making the hypermeter grid for Rnadom forest: 
parameter_ran_for = {'model__n_estimators': [100, 200, 300],'model__max_depth': [None, 10, 20],'model__min_samples_split': [2, 5, 10],'model__min_samples_leaf': [1, 2, 4]}

# Note: again Set up GridSearchCV:
rf_grid = GridSearchCV(ran_for_pipeline, parameter_ran_for, cv=kfolds,scoring="neg_root_mean_squared_error", n_jobs=-1)

# Note: we would fit the model using the CV to fin best hyperparameters
rf_grid.fit(X_train, y_train)

print("Best Random Forest hyperparameters:", rf_grid.best_params_)

# Note: here again doing the predictions on the test data using the best model
y_pred = rf_grid.predict(X_test)

# Note: Compute performance metrics: RMSE and R^2
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

# Output results
print("Random Forest Regression Results:")
print("RMSE:", rmse)
print("R^2:", r2)




Best Random Forest hyperparameters: {'model__max_depth': None, 'model__min_samples_leaf': 2, 'model__min_samples_split': 2, 'model__n_estimators': 300}
Random Forest Regression Results:
RMSE: 266097.0060834489
R^2: 0.8217393076182634


**Write-up!** Why does a Random Forest ensure better results than a single Decision Tree? 

**Your Answer Here:** (feel free to make multiple cells!)
A single decision tree is known to be highly prone to overfitting it has a tendency to memorize everything in the data also the noise and the outliers. Which makes predictions that can fall apart when face with new information. Random Forests, on the other hand, are way smarter. Instead of relying on one tree's opinion, they build an entire forest of trees(multiple trees) each tree is trained in a slightly different random sample of the data. This is why its so good all the different trees are evaluated and the averages their predictions  which leads to cancellation of each individual trees mistakes. The overfitting in single trees are wiped out and produces a more smoother model. Its also not just about building different various trees but random forest uses this technique called bagging or called bootstrap aggregating where each tree is trained on random subset of the data, which means  not two trees will be completely the same which leeds to a more stable average. A single decision tree is prone to high variance. Also at each split random forrest only considers random subset of features instead of all features which helps prevent single strong predictor and also encourages tree diversity.

## 3. Feature Extraction and Selection

### Problem 3.1: Feature Extraction

Apply **feature extraction** in this section in order to derive at least 2 new features. Be sure to apply the same feature extraction to both the training and testing datasets.

**Do this!** In the cells below, **explain** and perform the steps for feature extraction. Make sure that your implementations and write-ups are presented well and effectively describe your workflow. You may add comments or markdown cells for your documentation as you see fit!  

*Grading Note: Your work will be graded for _readability_, _style_, and _cleanlines_. So, carefully document your code and use descriptive/intuitive variable names. For write-ups use consice and clear language like in a written project report.*

> **Hint**: You can use our previous labs as examples of how you might do this. Remember, you may add as many cells (for code and text) as you need below. Here — we gave you one for free!

In [14]:
# your code here

# Note: first we make copies of the original X_train and X_test to avoid messing with the origincal data
X_train_for_feature_ex = X_train.copy()
X_test_for_feature_ex = X_test.copy()


# Note : we would like to use Feature 1 as  HouseAge
# The Reason are as follows :
#    YearBuilt has many missing values and are not directly usefull here
#    HouseAge = 2018 - YearBuilt (since dataset ends in 2018).
#    Older homes may be cheaper or renovated, adding predictive signal.


# Note: Now we fill missing YearBuilt values with our median just inthe training set\
yearbuilt_median = X_train_for_feature_ex['YearBuilt'].median()

X_train_for_feature_ex['HouseAge'] = 2018 - X_train_for_feature_ex['YearBuilt'].fillna(yearbuilt_median)
X_test_for_feature_ex ['HouseAge'] = 2018 - X_test_for_feature_ex ['YearBuilt'].fillna(yearbuilt_median)



# Note: Feature 2 - BathroomPerRoom
# The reasons are as follows:
#   Homes with very few bathrooms relative to their number of rooms may be less desirable.
#   This ratio captures bathroom availability per room, a useful measure of livability.
#   It also converts two features into one meaningful interaction feature.

# Note: Avoid division by zero by replacing Rooms == 0 with 1
X_train_for_feature_ex['BathroomPerRoom'] = X_train_for_feature_ex['Bathroom'] / X_train_for_feature_ex['Rooms'].replace(0, 1)
X_test_for_feature_ex['BathroomPerRoom'] = X_test_for_feature_ex['Bathroom'] / X_test_for_feature_ex['Rooms'].replace(0, 1)

# Print changes
print("New features added: 'HouseAge' and 'BathroomPerRoom'")
print(X_train_for_feature_ex[['HouseAge', 'BathroomPerRoom']].head())



New features added: 'HouseAge' and 'BathroomPerRoom'
       HouseAge  BathroomPerRoom
12796      48.0         0.500000
9642       48.0         0.333333
3207       48.0         0.500000
1698       43.0         0.500000
761        48.0         0.500000


**Write up!** Explain the feature extraction methods have you implemented and why.
The raw data set sometimes is fine but if you want better predictions then you have to get creative. This is wehre we bring in feature engineering which is taking all the features we have and transorming it into something that is way more useful. We have created two new features houseAge and bathroom. These are not random I think we can use these features to help our model in some way. We would definitely think that there is some relation to houseage and price We also think this would be some linear relation. Also bathrrom per bedroom is also something I personally would look into before buying a house, it is always better to have a greater ratio on it.

Let's begin with HouseAge. There is a YearBuilt column in the dataset, but hoenstly in its raw state, it is rather pointless. It has a lot of missing numbers, to start, and what is a model supposed to do with "1970" versus "2005"? The age of the house is more important than the year itself. Because the dataset expires in 2018, We turned it into an age-based feature: HouseAge = 2018 - YearBuilt. The model can now plainly observe that modern residences typically command higher costs, while older properties may be less expensive since they require maintenance. To keep things tidy and prevent any covert data leaking between train and test sets, We used the training set median to fill in the missing YearBuilt values. 

Now for the  second feature, BathroomPerRoom, it is  designed to capture how balanced and functional a home’s layout is. While the dataset includes the raw counts of rooms and bathrooms, these values alone don’t fully describe how livable a property might be. For example, a home with six rooms but only one bathroom may feel inconvenient, whereas a property with four rooms and two bathrooms offers a far more practical setup. By taking the ratio of bathrooms to rooms, this feature highlights whether a home has adequate bathroom availability relative to its size.

**Your Answer Here:**

### Problem 3.2: Feature Selection

Using the original and extracted features as input, perform **feature selection** in this section (statistical, model-based, or iterative feature selection). Be sure to apply the same feature selection to both the training and testing datasets. Note that categorical variables will need to be encoded and missing values should be appropriately delt with using imputation.

**Do this!** In the cells below, **explain** and perform the steps necessary for feature selection. Make sure that your implementations and write-ups are presented well and effectively describe your workflow. You may add comments or markdown cells for your documentation as you see fit!  

*Grading Note: Your work will be graded for _readability_, _style_, and _cleanlines_. So, carefully document your code and use descriptive/intuitive variable names. For write-ups use consice and clear language like in a written project report.*

> **Hint**: You can use our previous labs as examples of how you might do this. Remember, you may add as many cells (for code and text) as you need below. Here — we gave you one for free!

In [15]:
# your code here
# Problem 3.2 — Feature Selection

from sklearn.feature_selection import SelectKBest, f_regression
import pandas as pd

# Note: We would make copies of the extracted feature datasets
X_train_for_feature_selection = X_train_for_feature_ex.copy()
X_test_for_feature_selection = X_test_for_feature_ex.copy()

# Note : We are Initializing SelectKBest with f_regression to pick our top features
selector = SelectKBest(score_func=f_regression, k=10)

# Note: Fit selector on training data 
selector.fit(X_train_for_feature_selection, y_train)

# Note: transforming both training and testing tests using our selcted features
X_test_selected = selector.transform(X_test_for_feature_selection)

# Note: We now identify the names of our selected features
selected_mask = selector.get_support()
selected_features = X_train_for_feature_selection.columns[selected_mask]


print("Selected Features (Top 10):")
print(selected_features)




ValueError: could not convert string to float: 'Mount Waverley'

**Write up** Explain which feature selection method have you selected and why.

**Your Answer Here:**

## Problem 4: Preprocessing and Feature Engineered Models

### Problem 4.1: Preprocessing for the Feature Engineered Models

Now that we've removed/added features during feature engineering, we need to update the preprocessing from Problem 2.1. 

In [ ]:
# your code here




### Problem 4.2: Perform Linear Regression

Perform Linear Regression on **engineered training data**. Predict the output for the test dataset using the fitted model. Print the root mean squared error (RMSE) and R^2 value from Linear Regression for test data set.


In [ ]:
# your code here




### Problem 4.3 : Perform K-Nearest Neighbors Regression

Perform K-Nearest Neighbors Algorithm for Regression using cross validation to tune any hyperparameters, such as `n_neighbors`. Be sure to use the **data with feature engineering applied**.

Then, using the optimal hyperparameters, train a K-Nearest Neighbors Regressor on the entirety of the training data and predict output for the test dataset using the fitted model.

Print root mean squared error (RMSE) and R^2 from K-Nearest Neighbors Regression on the test datset.

In [ ]:
# your code here




### Problem 4.4 : Perform Decision Tree Regression 

Perform Decision Tree Regression using cross validation to tune any hyperparameters, such as `max_depth`. Be sure to use the **data with feature engineering applied.**

Then, using the optimal hyperparameters, train a Decision Tree Regressor on the entirety of the training data and predict output for the test dataset using the fitted model.

Print root mean squared error (RMSE) and R^2 from Decision Tree Regression on the test dataset.

In [ ]:
# your code here




### Problem 4.5 : Perform Random Forest Regression

Perform Random Forest Regression using cross validation to tune any hyperparameters, such as `n_estimators`. Be sure to use the **data with feature engineering applied.**

Then, using the optimal hyperparameters, train a Random Forest Regressor on the entirety of the training data and predict output for the test dataset using the fitted model.

Print root mean squared error (RMSE) and R^2 from Random Forest Regression on the test dataset.

In [ ]:
# your code here




## 5. Model Comparision

### Problem 5.1:

Create some visualizations that compare all the models results using either $R^{2}$ or RMSE. Be sure to include the following visualizations:
1. Performance comparison across the 4 baseline models from problem 2.
2. Performance comparison across the 4 feature engineered models from problem 4.
3. Performance comparison of the baseline models and the feature engineered models.

Feel free to include additional visualizations!

In [ ]:
# your code here




### Problem 5.2:

**Write up** Summarize your results and provide some crucial insights about the entire with and without feature engineering process such as which model performed the best and why, based on what metrics and indicators, etc. Which model benefited the most/the least from the feature engineering (if any)?

**Answer here:** (feel free to make multiple cells!)

That's it! Follow the submission instructions below:
* Restart your kernel and run all the cells
* Save your notebook
* Upload the .ipynb notebook directly to Gradescope

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [ ]:
grader.check_all()

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

In [ ]:
# Save your notebook first, then run this cell to export your submission.
grader.export()